Fullscreen Mode: Change the "Back" button to toggle the sidebar

shaunandrews opened the pull request:
With the recent move to make Fullscreen mode the default, I've heard from many people that they expect to be able to toggle the WordPress sidebar menu by clicking on the WordPress logo in the top left. This PR tries that:

![Fullscreen Sidebar Toggle i1](https://user-images.githubusercontent.com/191598/77475871-d8cac480-6def-11ea-8e2c-bc1115d7e6d4.gif)

shaunandrews committed:
A quick and dirty try at making the wp-admin sidebar accessible while in fullscreen-mode.

shaunandrews committed:
A quick and dirty try at making the wp-admin sidebar accessible while in fullscreen-mode.

shaunandrews committed:
Back to the dark background by default, and accounting for the `folded` state of the sidebar.

MichaelArestad commented:
I want this and I want it now! <3

mapk commented:
Moving this functionality to the top level is great. While it doesn't provide the top bar, it's still usable and surfaces the wp-admin navigation with ONE click. Great exploration!

SchneiderSam commented:
This is soooooo cool.

paaljoachim commented:
This is awesome! It sure makes going into Full Screen mode a lot easier.

Along with this PR let's also get a shortcut in place. https://github.com/WordPress/gutenberg/issues/20626

shaunandrews committed:
Making the WP logo just a little smaller.

shaunandrews commented:
A couple of things I could use help with:

- I think when you toggle the sidebar open, the keyboard `:focus` should jump to the wp-admin sidebar.
- Right now, the sidebar sticks around until you click the WP logo again. Part of me thinks that clicking anywhere _outside_ of the sidebar should hide the sidebar again. In fullscreen-mode, I don't think the sidebar should stick around.

I'm not sure how to go about implementing either of the above, but would love to learn. All tips, tricks, and commits welcomed.

paaljoachim commented:
Testing out icons for bringing back the sidebar menu.

Original.
![Sidebar-W-original](https://user-images.githubusercontent.com/5323259/77640328-7cd07f00-6f5a-11ea-9ff0-6588f29830fa.jpg)

Grey W.
![Sidebar-W-Gutenberg-5](https://user-images.githubusercontent.com/5323259/77640377-907be580-6f5a-11ea-8464-b55d7a0f9739.jpg)

White W.
![Sidebar-W-Gutenberg-7](https://user-images.githubusercontent.com/5323259/77640601-f9fbf400-6f5a-11ea-9a0a-78fed7c5046d.jpg)

Grey W with vertical line.
![Sidebar-grey-W-with-line](https://user-images.githubusercontent.com/5323259/77640772-40515300-6f5b-11ea-8e79-555f6bf19e6a.jpg)

Grey W with dashes and vertical line.
![Sidebar-W-Gutenberg-1](https://user-images.githubusercontent.com/5323259/77640845-58c16d80-6f5b-11ea-86c5-a084e29c2e1f.jpg)

White W with dashes and vertical line.
![W-show-sidebar-Gutenberg-2](https://user-images.githubusercontent.com/5323259/77640917-7393e200-6f5b-11ea-9b77-1c949a729dfe.jpg)

White W with dashes and vertical line 2.
![Show-sidebar-W-Gutenberg-3](https://user-images.githubusercontent.com/5323259/77640962-84dcee80-6f5b-11ea-896d-e742fd64be8a.jpg)

Grey W with horizontal dashes.
![Sidebar-W-Gutenberg-4](https://user-images.githubusercontent.com/5323259/77641024-a50cad80-6f5b-11ea-870a-c3984f8d4a45.jpg)

bph commented:
This is an awesome idea, and well executed. Thank you @shaunandrews :-) Sorry, I can't help you with development but I wanted you to know that a lot of end users will be quite happy with this.

mtias commented:
Thanks for exploring this @shaunandrews. We should look at the centered logo a bit in the open state: it'd make sense to see if we can render the site title when it expands (and keep it hidden when collapsed) to balance it better.

ItsJonQ commented:
> A couple of things I could use help with:
>
> * I think when you toggle the sidebar open, the keyboard :focus should jump to the wp-admin sidebar.
>
> * Right now, the sidebar sticks around until you click the WP logo again. Part of me thinks that clicking anywhere outside of the sidebar should hide the sidebar again. In fullscreen-mode, I don't think the sidebar should stick around.

@shaunandrews Haii!! Great work on this update! I can certainly help you with these items :)

re: Focus jump:

To do this, we'll need to hook into some sort of `onOpen` state transition to make the focus. I think maybe keeping the focus on the "W" logo may be better. However, perhaps when...

* Focus is on "W" logo, and...
* Menu is open, and...
* User presses `DOWN` on the keyboard,...
* `:focus` jumps to the first menu item

re: Closing menu when clicking outside

This is often accomplished using 2 techniques:

1. "Overlay"

When the menu is open, an "Overlay" (a transparent `div`) is rendered on top of the Editor, example:

![overlay](https://user-images.githubusercontent.com/2322354/78032360-e4a81080-7332-11ea-874d-f79d253c4394.png)

This *blocks* interaction for anything underneath it, and allows you to easily attach an `onClick` handler on it. Which means, when you click on the overlay, the menu can close.

2. "Clicking outside"

This is the trickier, but often more preferrable, way to handle these interactions. What needs to be done is a `click` event listener needs to be added, and logic needs to read something like:

* If the (clicked) `event.target` is...
* **NOT** the "W" logo, and...
* **NOT** the menu, and...
* **NOT** the top toolbar,....
* Then close.

Hope this helps 🙏

mapk commented:
This is beautiful! 
I'm in favor of seeing a prototype that works in Matias' suggestion. Keep the logo fixed to the left and add the site title when expanded. 

This is also the desire after [today's triage meeting in slack](https://wordpress.slack.com/archives/C02S78ZAL/p1585670695090100).

ItsJonQ commented:
Haiii!! I coordinated with @shaunandrews .
I'm gonna try to help out a bit with the JS code side.

🤞

Commit:
Improve interaction handling for Navigation menu logo

Create custom React hook to encapsulate state, click, focus, and keyboard logic.

ItsJonQ commented:
Heya @shaunandrews !

I have updates :). I've made some refactors and enhancements in a separate branch:
https://github.com/WordPress/gutenberg/tree/try/fullscreen-sidebar-toggle-refactor

(This branch was also rebased with latest `master`)

You can see my changes here:
https://github.com/WordPress/gutenberg/commit/9a9f9efeffdaeb2d71d0a00d4798d63a67a9f44b

TLDR:
* Create a custom React hook to better manage the open/close menu state
* Leveraged hook to bind the click + keyboard events for closing or (re)focusing

Let me know what you think!

---

Note: I haven't done anything with Site Title yet.

> it'd make sense to see if we can render the site title when it expands (and keep it hidden when collapsed)

We can do this. However, I'm not too sure what it may look like if the site title was **longer**. In the regular WP Admin, the "W" logo and title are separate from the menu, allowing longer titles to flow.

I think a mockup of this would be helpful 😊

Thank you!

Commit:
Adjust click outside to close logic

Commit:
Fix SVG color and pointer events

shaunandrews commented:
> Let me know what you think!

Super cool, @ItsJonQ — thanks for your help on this. I'd love to see your branch merged into this one.

One thing I'm still unsure about is the keyboard interaction. Having to press the down arrow to start navigating the wp-admin sidebar seems strange. If you look at how the inserter works, when you tab to it and select, the `:focus` goes into the popover. I feel like this maps to the wp-admin toggle; Selecting the logo with the keyboard seems like it should move the `:focus` into the menu.

--

As for the site title, I like the idea in theory. But I'm not sure there's enough room for it with the current width of the menu. Here's what it would look like with a truncated site title:

<img width="502" alt="image" src="https://user-images.githubusercontent.com/191598/78162986-ac2d3300-7415-11ea-91a8-d72bbc059a9d.png">

We _could_ change the width of the menu (only within the Editor context) to add some overall space, but I'd expect that to feel strange in use:

<img width="605" alt="image" src="https://user-images.githubusercontent.com/191598/78163126-ebf41a80-7415-11ea-979b-fc97d475f652.png">

We could break the title down to its own line:

<img width="460" alt="image" src="https://user-images.githubusercontent.com/191598/78163324-3ffeff00-7416-11ea-9a2a-f5d67d5bee2e.png">

Commit:
Rename fullscreen-mode-close to admin-menu-toggle

ItsJonQ commented:
@shaunandrews Thanks for 🍐 'ing with me on this one!

I just rebased my changes from my `refactor` branch onto yours and pushed it up.

One thing we noticed was that the name of `FullscreenModeClose` no longer described what the component did.

I renamed it to `AdminMenuToggle`, and have adjusted CSS classNames and E2E tests accordingly.

Commit:
Fix classNames and imports for renamed admin-menu-toggle

Commit:
Focus Admin Menu on TAB press when open

ItsJonQ commented:
> Selecting the logo with the keyboard seems like it should move the :focus into the menu.

@shaunandrews Ah yes, the keyboard interactions are tricky for this one. I just pushed an update that focuses using `TAB` rather than `DOWN`.

Let me know how this feels :)

enriquesanchez commented:
Keyboard interaction feels good to me, but I found one small issue.

If I move focus to the "W" and hit Enter, the wp-admin menu expands and my focus stays on the "W". This is expected. I can then move focus to the first item in the menu by hitting `TAB`.

The one problem I found was going back. If try to do `shift + TAB` to go back up to the "W" button, I can't seem to be able to land focus on it. I do see a "Skip to toolbar" skip link, which then takes me to the "W" button if I do one more `TAB`, but that was unexpected. I expected to be able to go back the same way I came down.

If I use the "Skip to toolbar" link, I'm not really taken to the "toolbar" because if I do a `TAB` again after focusing on the "W", my focus moves down to the "Dashboard" in the menu. That is unexpected, I assumed that because I'm now on the toolbar I would move focus to the inserter button instead.

I would need to collapse the menu first, so that I'm now on the toolbar and can move to the inserter.

shaunandrews commented:
This breaks the top toolbar:

<img width="483" alt="image" src="https://user-images.githubusercontent.com/191598/78177198-bce8a380-742b-11ea-9177-68340dc6afca.png">

timhibberd commented:
@shaunandrews I like this a lot but this would tie a navigation function to an icon that a lot of MSPs hide for business sites. So the question is...if this is hidden can we be assured of no side-effects? Will the icon that inherits that position now inherit the navigation logic?

If you are going to do this, anchor it as a mandatory navigation element with a php API to replace it with a different icon / function. Even that would break on legacy sites who don't know about the new API.

noahtallen commented:
@timhibberd It looks like this is planned as part of https://github.com/WordPress/gutenberg/issues/20929

timhibberd commented:
Excellent. Love it (combo of this PR + #20929 )  :-)

Clorith commented:
My initial thought is to not move the WP logo, this is supposed to bridge the gap between "losing" the sidebar, and it should retain the same visual appearance of the sidebar then.

shaunandrews committed:
Make sure the top toolbar moves when the menu is opened.

noahtallen commented on packages/block-editor/src/components/fullscreen-mode/style.scss:
```diff
@@ -23,7 +22,6 @@ body.js.is-fullscreen-mode {
 			transition: left 0.12s ease-in-out;
 		}
 
-		// This is all so stupid
```
🤣

shaunandrews replied:
lol, I often make a lot of comments as I'm working on code. Most get deleted. They usually include much more... harsh... language. 😆

ItsJonQ replied:
😂

shaunandrews committed:
Adjusting focus state to use new line-width, making the button a little bigger, and only changing the fill color on hover.

tellthemachines commented on packages/block-editor/src/components/fullscreen-mode/style.scss:
```diff
@@ -10,27 +9,69 @@ body.js.is-fullscreen-mode {
 		}
 
-		#adminmenumain,
+		// Hide the top adminbar toolbar.
```
We might need to add and remove the `display: none` with JS (as adding it directly in the CSS might ruin the animation). The problem is that currently all the nav elements, and the skip links, are still focusable when hidden, and that causes some weirdness when tabbing through them:

<img width="219" alt="Screen Shot 2020-04-03 at 4 25 08 pm" src="https://user-images.githubusercontent.com/8096000/78327463-a9267400-75c8-11ea-832c-c0dc66d86e3d.png">

<img width="181" alt="Screen Shot 2020-04-03 at 4 24 54 pm" src="https://user-images.githubusercontent.com/8096000/78327471-af1c5500-75c8-11ea-814c-f2ee9a2cc5d2.png">

shaunandrews replied:
Ah, good catch. I _think_ adding `pointer-events: none;` might work here. Using `display: none;`, like you said, will break the animation. I'll see what I can do to resolve this — thanks for taking a look.

ItsJonQ replied:
@tellthemachines + @shaunandrews Haiii! I just pushed something up that resolves this (I believe).

It hides the element wrapping the menu items. This allows for the outer (bar) to retain it's animation transition properties, but also prevents the inner links from being tabbable.

tellthemachines reviewed (commented):
Looks great overall, just one small issue below 🙂

Commit:
Improve admin menu keyboard navigation in open/close states

vindl commented on packages/interface/src/components/fullscreen-mode/style.scss:
```diff
@@ -10,14 +9,80 @@ body.js.is-fullscreen-mode {
+	}
+
+	.edit-post-header-toolbar__block-toolbar,
+	.block-editor-editor-skeleton__body,
```
I renamed this to `InterfaceSkeleton` in https://github.com/WordPress/gutenberg/pull/21335, so the classnames here will need to be updated.

shaunandrews replied:
Thanks for the heads up. I've updated the classnames.

vindl commented on packages/edit-post/src/components/header/admin-menu-toggle/index.js:
```diff
@@ -0,0 +1,128 @@
+import { __ } from '@wordpress/i18n';
+import { wordpress } from '@wordpress/icons';
+
+function AdminMenuToggle() {
```
Since we'll need the same component in Site Editor and maybe Widgets screen, could we move it to `@wordpress/interface` packaged so it can be more easily reused in that context?

shaunandrews replied:
Cool, I managed to figure out how to move it to the interface package. Should be good to go now.

paaljoachim commented:
Which shortcut should we use for going in and out of Fullscreen mode?
I am asking as I have an issue in relation to it: https://github.com/WordPress/gutenberg/issues/20626

shaunandrews committed:
Updating classNames after a rebase.

shaunandrews committed:
Moving to toggle button from edit-post into the interface package so it can be used elsewhere.

youknowriad commented on packages/edit-post/src/components/header/index.js:
```diff
@@ -6,12 +6,12 @@ import { Button } from '@wordpress/components';
  * Internal dependencies
  */
-import FullscreenModeClose from './fullscreen-mode-close';
+//import AdminMenuToggle from './admin-menu-toggle';
```
I guess this should be removed

shaunandrews replied:
Yup. Gone.

youknowriad commented on packages/interface/src/components/admin-menu-toggle/index.js:
```diff
@@ -0,0 +1,128 @@
+
+	const { isActive, postType } = useSelect( ( select ) => {
+		const { getCurrentPostType } = select( 'core/editor' );
+		const { isFeatureActive } = select( 'core/edit-post' );
```
This component has been moved to a generic package "interface" which means it can't use a specific store like "core/edit-post" or "core/editor". I'm not even certain why we're checking here the post type here. For me we should remove all that use select code and the check to return null a bit under.

We should just not render this component from its parent when we're not on fullscreen.

shaunandrews replied:
Took me a bit to understand this, but now it makes sense. I've removed the display logic out of the component and instead moved it into where it's used.

youknowriad commented on packages/interface/src/components/admin-menu-toggle/index.js:
```diff
@@ -0,0 +1,128 @@
+
+	return (
+		<Button
+			className="edit-post-fullscreen-admin-menu-toggle"
```
This classname should probably be updated to reflect the new package.

shaunandrews replied:
Nice catch. Updated.

youknowriad commented on packages/interface/src/components/admin-menu-toggle/style.scss:
```diff
@@ -0,0 +1,53 @@
+			}
+		}
+
+		body.js.is-showing-admin-menu & {
```
Why do we need the `js` className on all these selectors?

shaunandrews replied:
I guess we don't? I'm not sure why I added it.

youknowriad commented on packages/interface/src/components/fullscreen-mode/style.scss:
```diff
@@ -10,14 +9,80 @@ body.js.is-fullscreen-mode {
+			top: 60px;
+		}
+
+		.edit-post-header-toolbar__block-toolbar,
```
We shouldn't be targeting edit-post styles on the interface package. What is this about?

youknowriad replied:
Looks like there's more of these styles a bit under. can't we target the skeleton header instead?

shaunandrews replied:
I see. That makes sense. I've updated it.

youknowriad commented on packages/edit-post/src/components/header/index.js:
```diff
@@ -58,7 +58,7 @@ function Header() {
 	return (
 		<div className="edit-post-header">
-			<FullscreenModeClose />
+			<AdminMenuToggle />
```
Can we do the same for the edit-site package (Site builder)? This will help you figure out better how to make these styles generic and work across screens.

youknowriad reviewed (commented):
Great work here.

shaunandrews committed:
Oops, forgot to remove this.

shaunandrews committed:
Updating classnames.

shaunandrews committed:
Replacing the FullscreenModeClose component with the new AdminMenuToggle.

shaunandrews committed:
Removing the .js selector.

mapk commented:
It's going to be interesting to see how this works in relation to https://github.com/WordPress/gutenberg/pull/20951. Any thoughts around how these two slide-outs work together, @shaunandrews?

shaunandrews committed:
Moving the display logic out of the component and into the places where its used.

shaunandrews commented:
> Any thoughts around how these two slide-outs work together

The wp-admin menu will simply push the inserter over, just like the rest of the editor's skeleton. I don't foresee any interference or issues.

ItsJonQ commented:
Looks like Travis is unhappy. Let me see if I can help! 💪

Commit:
Update package-lock and e2e test

shaunandrews commented:
> Any thoughts around how these two slide-outs work together

The wp-admin menu will simply push the inserter over, just like the rest of the editor's skeleton. I don't foresee any interference or issues.

joyously commented on packages/edit-post/src/components/header/index.js:
```diff
@@ -6,12 +6,11 @@ import { Button } from '@wordpress/components';
 import { useSelect, useDispatch } from '@wordpress/data';
 import { cog } from '@wordpress/icons';
-import { PinnedItems } from '@wordpress/interface';
+import { PinnedItems, AdminMenuToggle } from '@wordpress/interface';
```
I know it's not a component, but the _is_ an existing thing that collapses the admin menu. I'm not sure what it's called, but it's at the bottom of the menu. I think using the name `AdminMenuToggle` could be a bit confusing since one already exists. I'm not sure why the existing one isn't used for this fullscreen thing.

shaunandrews replied:
I guess I think of the `#collapse-button` option that exists at the bottom of the menu as a different feature; The collapse button changes the width of the admin sidebar, not the visibility.

This new `AdminMenuToggle` component toggles the visibility.

The two features can work together; You can collapse the admin sidebar _and_ toggle its visibility.

gziolo commented on packages/interface/src/components/admin-menu-toggle/index.js:
```diff
@@ -0,0 +1,112 @@
+			}
+		};
+
+		if ( isActive ) {
```
How about you move the check for `isActive` to `handleOnKeyDown` callback to make the setup of event handlers independent from React state changes?

ItsJonQ replied:
Oo! That works for me :) Thank you 🙏

youknowriad commented on packages/interface/src/components/fullscreen-mode/style.scss:
```diff
@@ -10,14 +9,65 @@ body.js.is-fullscreen-mode {
+		.interface-interface-skeleton__body,
+		.interface-interface-skeleton__footer {
+			transition: margin-left 0.12s ease-in-out;
+		}
```
All these `skeleton-*` styles on this stylesheet make the "full screen mode" work properly if your screen is using the Skeleton. do you think it's possible to make it independent of that? I mean no matter how the current screen is rendered, it could work?

I don't see this as a blocker but more "good to have"

youknowriad reviewed (approved):
on Safari, the wp menu doesn't cover the full height of the window.  (see the white space at the bottom)

<img width="165" alt="Capture d’écran 2020-04-15 à 9 49 13 AM" src="https://user-images.githubusercontent.com/272444/79318082-79e40100-7efe-11ea-94ce-a5e6f8131896.png">

Other than that, I think this is looking good.

Commit:
Refactor event bindings in relation to state within admin-menu-toggle

shaunandrews committed:
Refactoring the transition to avoid a display bug in Safari.

shaunandrews commented:
Rebased and fixed the display bug in Safari.

Commit:
Update package-lock.json

jasmussen commented:
The left menu doesn't scroll when in fullscreen mode on a small viewport:

<img width="1165" alt="Screenshot 2020-04-24 at 14 58 32" src="https://user-images.githubusercontent.com/1204802/80215295-7eb25e80-863c-11ea-8aae-013baba7b208.png">

The "fix" in non-fullscreen mode is to allow the HTML element to scroll, causing a double scrollbar on the right. Not an ideal fix, and would love ideas for making that better.

mtias commented:
I think we need to revert this one and continue working on it for a bit. There are aspects that need to be better handled for an adequate integration, otherwise states like this are possible:

![image](https://user-images.githubusercontent.com/548849/80394855-bb878b00-88b2-11ea-82b7-0b898536de59.png)

aduth commented:
Revert pull request at #21929
